language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddOnFailCodeAction.java

```diff
@@ -0,0 +1,122 @@
+                .equals(diagnostic.diagnosticInfo().code())) &&
+                !(DiagnosticErrorCode.FAIL_EXPR_NO_MATCHING_ERROR_RETURN_IN_ENCL_INVOKABLE.diagnosticId()
+                        .equals(diagnostic.diagnosticInfo().code()))) {
+            return Collections.emptyList();
+        }
+
+        Node nodeAtDiagnostic = positionDetails.matchedNode();
+        List<TextEdit> edits = new ArrayList<>();
+        String commandTitle;
+
+        FailStatementResolver finder = new FailStatementResolver(diagnostic);
+        Optional<Node> failStatementResolverNode = finder.getRegularCompoundStatementNode(nodeAtDiagnostic);
+        
+        if (failStatementResolverNode.isPresent()
+                && failStatementResolverNode.get().kind() == SyntaxKind.WHILE_KEYWORD) {
+            while (nodeAtDiagnostic.kind() == SyntaxKind.WHILE_STATEMENT) {
+                nodeAtDiagnostic = nodeAtDiagnostic.parent();
+            }
+            edits.add(getSurroundWithOnFailEditText(nodeAtDiagnostic));
+            commandTitle = CommandConstants.SURROUND_WITH_DO_ON_FAIL;
+            return Collections.singletonList(createQuickFixCodeAction(commandTitle, edits, context.fileUri()));
+        }
+        
+        if (failStatementResolverNode.isPresent()) {
+            LinePosition regularCompoundStmtLinePosition = failStatementResolverNode.get().lineRange().endLine();
+            Position onFailPosLine = new Position(regularCompoundStmtLinePosition.line(), 
+                    regularCompoundStmtLinePosition.offset());
+
+            String spaces = " ".repeat(regularCompoundStmtLinePosition.offset() - 1);
+            String editText = " on fail var e {" + CommonUtil.LINE_SEPARATOR + spaces + "\t"
+                    + CommonUtil.LINE_SEPARATOR + spaces + "}";
+            edits.add(new TextEdit(new Range(onFailPosLine, onFailPosLine), editText));
+            commandTitle = CommandConstants.CREATE_ON_FAIL_CLAUSE;
+        } else {
+            edits.add(getSurroundWithOnFailEditText(nodeAtDiagnostic));
+            commandTitle = CommandConstants.SURROUND_WITH_DO_ON_FAIL;
+        }
+
+        return Collections.singletonList(createQuickFixCodeAction(commandTitle, edits, context.fileUri()));
+    }
+ 
+    public TextEdit getSurroundWithOnFailEditText(Node node) {
+        while (!StatementNode.class.isAssignableFrom(node.getClass())) {
+            node = node.parent();
+        }
+        LinePosition startLinePosition = node.lineRange().startLine();
+        LinePosition endLinePosition = node.lineRange().endLine();
+        Position positionDo = new Position(startLinePosition.line(), startLinePosition.offset());
+        Position posCheckLineStart = new Position(endLinePosition.line(), endLinePosition.offset());
+
+        String spaces = " ".repeat(startLinePosition.offset());
+        String insideDoStatement = node.toSourceCode();
+        String tabShiftedDoStatement = insideDoStatement.substring(0, insideDoStatement.length() - 1)
```

Comment:
Why do we need `length()-1` here? Won't the last `\n` be replaced as well?

Reply:
The first line gets its `\t` from the hard coded "\t"` in line `110` and the last line gets an extra `\t` which has been removed by the logic in line `108`